cloud/google/cmd/gce-machine-controller/app/controller.go

```diff
@@ -0,0 +1,156 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package app
+
+import (
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/kubernetes-incubator/apiserver-builder/pkg/controller"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	"k8s.io/client-go/kubernetes"
+	v1core "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/leaderelection"
+	"k8s.io/client-go/tools/leaderelection/resourcelock"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/cluster-api/cloud/google"
+	"sigs.k8s.io/cluster-api/cloud/google/cmd/gce-machine-controller/app/options"
+	"sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset"
+	clusterapiclientsetscheme "sigs.k8s.io/cluster-api/pkg/client/clientset_generated/clientset/scheme"
+	"sigs.k8s.io/cluster-api/pkg/controller/config"
+	"sigs.k8s.io/cluster-api/pkg/controller/machine"
+	"sigs.k8s.io/cluster-api/pkg/controller/sharedinformers"
+)
+
+const (
+	gceMachineControllerName = "gce-machine-controller"
+)
+
+func StartMachineController(server *options.MachineControllerServer, shutdown <-chan struct{}) {
+	config, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Fatalf("Could not create Config for talking to the apiserver: %v", err)
+	}
+
+	client, err := clientset.NewForConfig(config)
+	if err != nil {
+		glog.Fatalf("Could not create client for talking to the apiserver: %v", err)
+	}
+
+	actuator, err := google.NewMachineActuator(server.KubeadmToken, client.ClusterV1alpha1().Machines(corev1.NamespaceDefault), server.MachineSetupConfigsPath)
+	if err != nil {
+		glog.Fatalf("Could not create Google machine actuator: %v", err)
+	}
+
+	si := sharedinformers.NewSharedInformers(config, shutdown)
+	// If this doesn't compile, the code generator probably
+	// overwrote the customized NewMachineController function.
+	c := machine.NewMachineController(config, si, actuator)
+	c.RunAsync(shutdown)
+
+	select {}
+}
+
+func Run(server *options.MachineControllerServer) error {
+	kubeConfig, err := controller.GetConfig(server.CommonConfig.Kubeconfig)
+	if err != nil {
+		glog.Errorf("Could not create Config for talking to the apiserver: %v", err)
+		return err
+	}
+
+	kubeClientControl, err := kubernetes.NewForConfig(
+		rest.AddUserAgent(kubeConfig, "machine-controller-manager"),
+	)
+	if err != nil {
+		glog.Errorf("Invalid API configuration for kubeconfig-control: %v", err)
+		return err
+	}
+
+	recorder, err := createRecorder(kubeClientControl)
+	if err != nil {
+		glog.Errorf("Could not create event recorder : %v", err)
+		return err
+	}
+
+	run := func(stop <-chan struct{}) {
+		StartMachineController(server, stop)
+		panic("unreachable")
```

Comment:
Why?

Reply:
Ah I see why now. The `Run` function returns an `error`, and `return nil` would be misleading. However this inner function doesn't have a return value, so the panic isn't needed.

Reply:
yeah - why the panic? That seems a bit odd - and the message seems equally as odd